server/src/main/java/org/elasticsearch/indices/SystemIndices.java

```diff
@@ -864,39 +903,61 @@ public MigrationCompletionHandler getPostMigrationFunction() {
          */
         public static void cleanUpFeature(
-            Collection<? extends IndexPatternMatcher> indexDescriptors,
+            Collection<SystemIndexDescriptor> indexDescriptors,
             Collection<? extends IndexPatternMatcher> associatedIndexDescriptors,
             String name,
             ClusterService clusterService,
             Client client,
-            ActionListener<ResetFeatureStateStatus> listener
+            final ActionListener<ResetFeatureStateStatus> listener
         ) {
             Metadata metadata = clusterService.state().getMetadata();
 
-            List<String> allIndices = Stream.concat(indexDescriptors.stream(), associatedIndexDescriptors.stream())
+            List<String> associatedIndices = associatedIndexDescriptors.stream()
                 .map(descriptor -> descriptor.getMatchingIndices(metadata))
                 .flatMap(List::stream)
                 .toList();
 
-            if (allIndices.isEmpty()) {
-                // if no actual indices match the pattern, we can stop here
+            final int taskCount = countInnerTasks(metadata, indexDescriptors, associatedIndices);
+
+            // check if there's nothing to do and take an early out
+            if (taskCount == 0) {
                 listener.onResponse(ResetFeatureStateStatus.success(name));
                 return;
             }
 
-            DeleteIndexRequest deleteIndexRequest = new DeleteIndexRequest();
-            deleteIndexRequest.indices(allIndices.toArray(Strings.EMPTY_ARRAY));
-            client.execute(DeleteIndexAction.INSTANCE, deleteIndexRequest, new ActionListener<>() {
-                @Override
-                public void onResponse(AcknowledgedResponse acknowledgedResponse) {
-                    listener.onResponse(ResetFeatureStateStatus.success(name));
-                }
+            GroupedActionListener<ResetFeatureStateStatus> groupedListener = new GroupedActionListener<>(
+                ActionListener.wrap(listenerResults -> {
+                    List<ResetFeatureStateStatus> errors = listenerResults.stream()
+                        .filter(status -> status.getStatus() == ResetFeatureStateResponse.ResetFeatureStateStatus.Status.FAILURE)
+                        .collect(Collectors.toList());
 
-                @Override
-                public void onFailure(Exception e) {
-                    listener.onResponse(ResetFeatureStateStatus.failure(name, e));
+                    if (errors.isEmpty()) {
+                        listener.onResponse(ResetFeatureStateStatus.success(name));
+                    } else {
+                        StringBuilder exceptions = new StringBuilder("[");
+                        exceptions.append(errors.stream().map(e -> e.getException().getMessage()).collect(Collectors.joining(", ")));
+                        exceptions.append(']');
+                        listener.onResponse(ResetFeatureStateStatus.failure(name, new Exception(exceptions.toString())));
```

Comment:
Could you also add logging that captures the full stack traces? It's nitpicky but it's no fun when we hit a weird error and realize there's no way to get the actual stack trace.

Reply:
Sure sounds good.